bait-files/src/test/java/com/google/sps/servlets/ReviewServletTest.java

```diff
@@ -155,4 +155,84 @@ public void doGet_whenReviewExistsAndRequestExists_returnsReviewAndRequest()
+    parameters.put("name-validity", new String[] {"valid"});
+    parameters.put("username-validity", new String[] {"valid"});
+    parameters.put("email-validity", new String[] {"invalid"});
+    parameters.put("phone-validity", new String[] {"invalid"});
+    parameters.put("address-validity", new String[] {"valid"});
+    parameters.put("image-validity", new String[] {"invalid"});
+    parameters.put("authenticity-rating", new String[] {"2"});
+    parameters.put("reviewer-notes", new String[] {"Look Like You Know - Royal Blood"});
+
+    when(request.getParameterMap()).thenReturn(parameters);
+    when(request.getParameter("status")).thenReturn("COMPLETED");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    new ReviewServlet().doPost(request, response);
+
+    Query<Review> query = ObjectifyService.ofy().load().type(Review.class);
+    List<Review> allReviews = query.list();
+
+    String rawJsonResponse = stringWriter.toString();
+    System.out.println(rawJsonResponse);
+
+    assertTrue(allReviews.size() == 1);
+  }
+
+  @Test
+  public void doPost_parameterMissing_sendRedirect() throws IOException, ServletException {
+    ObjectifyService.ofy()
+        .save()
+        .entity(new Request(14L, "4", "human", "human47", "human47@gmail.com", "2930 pearl street",
+            "no_image", "555-555-5555", "some notes"))
+        .now();
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+    Map<String, String[]> parameters = new HashMap<String, String[]>();
+    parameters.put("review-request-id", new String[] {"14"});
+    parameters.put("review-user-id", new String[] {"4"});
+    parameters.put("name-validity", new String[] {"valid"});
+    parameters.put("username-validity", new String[] {"valid"});
+    parameters.put("email-validity", new String[] {"invalid"});
+    parameters.put("phone-validity", new String[] {"invalid"});
+    parameters.put("address-validity", new String[] {""}); // missing parameter value
+    parameters.put("image-validity", new String[] {"invalid"});
+    parameters.put("authenticity-rating", new String[] {"2"});
+    parameters.put("reviewer-notes", new String[] {"Look Like You Know - Royal Blood"});
+
+    when(request.getParameterMap()).thenReturn(parameters);
+    when(request.getParameter("status")).thenReturn("COMPLETED");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    new ReviewServlet().doPost(request, response);
+
+    Query<Review> query = ObjectifyService.ofy().load().type(Review.class);
+    List<Review> allReviews = query.list();
+
+    String rawJsonResponse = stringWriter.toString();
+    System.out.println(rawJsonResponse);
+
+    assertTrue(allReviews.size() == 0);
```

Comment:
Can you also verify that the response.sendRedirect was called?
I think the syntax would be something like:
Mockito.verify(response).sendRedirect("[expected redirect path]");

Reply:
The expected behaviour here is to send a 400 error. Added a verification statement.